Make ds2 run on PaddleCloud and optimize performance

wanghaoshuang opened the pull request:
1. Refine data_utils/data.py to read bytes from tar file
2. Add scripts to submit paddle cloud job for ds2 trainning

Steps:
1. Pack train and test data according user's manifest file
2. Upload packed data and manifest file to PaddleCloud
3. Submit PaddleCloud job

fix #143

wanghaoshuang committed:
Make ds2 run on paddle cloud
1. Refine data_utils/data.py to read bytes from tar file
2. Add scripts to submit paddle cloud job for ds2 trainning

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -8,13 +8,20 @@
 import multiprocessing
+from threading import local
 import paddle.v2 as paddle
+import tarfile
```
Put `import tarfile` before `import paddle`. Widely used module comes first.

wanghaoshuang replied:
Fix.

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -64,7 +68,6 @@ def __init__(self,
                  window_ms=20.0,
                  max_freq=None,
                  specgram_type='linear',
-                 use_dB_normalization=True,
```
Why to remove `use_dB_normalization` ?

wanghaoshuang replied:
I didn't mean to remove `use_dB_normalization`. Maybe i made a mistake on git merging?

wanghaoshuang replied:
Fix.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+import os
```
Add a simple file doc, like other files.

xinghai-sun replied:
Move pcloud_split_data.py and pcloud_prepare_data.py (rename pclound_data.py to this) into deep_speech_2/cloud.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+import argparse
+
+
+def split_data(inManifest, tar_path, outManifest):
```
inManifest --> in_manifest,  outManifest --> out_manifest.
Keep consistency for variable naming style.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+def split_data(inManifest, tar_path, outManifest):
+    trainer_id = 1
+    trainer_count = 2
+    #with open("/trainer_id", "r") as f:
```
Remove L9-L12 or Remove L7-L8 ?
Please remove experimental or temporary code in the pull request.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+    #with open("/trainer_count", "r") as f:
+    #    trainer_count = int(f.readline()[:-1])
+
+    tarPath = os.path.abspath(tar_path)
```
tarPath --> tar_path

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/datasets/librispeech/pcloud_data.py:
```diff
@@ -0,0 +1,51 @@
+import json
```
Add a simple file doc.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/datasets/librispeech/pcloud_data.py:
```diff
@@ -0,0 +1,51 @@
+    manifest = []
+    for json_line in open(manifest_path):
+        try:
+            json_data = json.loads(json_line)
```
Could you reuse `read_manifest` in data_utils/utils.py ?

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description=__doc__)
+
+    parser.add_argument(
```
Let's keep consistent by putting `parser.add_argument` at the top of file.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/datasets/librispeech/pcloud_data.py:
```diff
@@ -0,0 +1,51 @@
+        sound_file = json_data['audio_filepath']
+        filename = os.path.basename(sound_file)
+        out_tar.add(sound_file, arcname=filename)
+        json_data['audio_filepath'] = filename
```
Once the cloud tar file is generated, the filepath in out_manifest should be modified to  "tar:%s#%s" immediately, instead of putting the  modification in pcloud_split_data.py.

wanghaoshuang replied:
When we generate tar file locally, we don't  known the real path of tar file on cloud. So  the  filepath in out_manifest can`t be modified to "tar:%s#%s" immediately.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+        help="Input manifest path. (default: %(default)s)")
+    parser.add_argument(
+        "--data_tar_path",
+        default='datasets/dev.tar',
```
Put the data in ./cloud

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_split_data.py:
```diff
@@ -0,0 +1,47 @@
+
+    parser.add_argument(
+        "--in_manifest_path",
+        default='datasets/dev.mani',
```
Put the file in ./cloud

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/pcloud_train.sh:
```diff
@@ -0,0 +1,32 @@
+#setted by user
```
Move pcloud_submit.sh into ./cloud

wanghaoshuang replied:
Fixed.

xinghai-sun replied:
Remove this file since it has been added in ./cloud

wanghaoshuang replied:
Ok. I forgot to delete this file.

xinghai-sun commented on deep_speech_2/pcloud_train.sh:
```diff
@@ -0,0 +1,32 @@
+#setted by user
+TRAIN_MANI='/pfs/dlnel/home/yanxu05@baidu.com/wanghaoshuang/data/ds2_data/demo.mani'
```
能否准备一个和单机版一样的配置，而不是使用demo数据？
并且注意将“wanghaoshuang” 等字符串替换掉，我们需要提供一个清晰的用户下载下来，少量改动后，就可以完整得run的版本。
另外，辛苦在README.md中仔细说明下cloud的使用。

wanghaoshuang replied:
@xinghai-sun 
Fixed.
1. 现在默认的数据与单机版一致。
2. 数据已转移到paddle cloud提供的公用路径下，无个人名字字样。
3. 在deep_speech_2/cloud/README.md中补充了使用公开数据跑ds2的说明。更多使用方式，我以后另起pr?

xinghai-sun reviewed (changes requested):
Great!
Could you please add a detailed tutorial for how to quickly start a paddle cloud version of DS2?

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/models into ds2_pcloud

wanghaoshuang committed:
Refine submitting scripts for deepspeech2 on paddle cloud.

wanghaoshuang committed:
Clean warning logs in cloud/README.md

wanghaoshuang committed:
Merge branch 'develop' into ds2_pcloud

xinghai-sun commented on deep_speech_2/cloud/pcloud_submit.sh:
```diff
@@ -0,0 +1,17 @@
+DS2_PATH=../
+tar -czf deepspeech.tar.gz ${DS2_PATH}
```
No need for this?

wanghaoshuang replied:
Yes. The latest paddlecloud client does't need this.

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,41 @@
+I0727 05:01:50.454787    25 GradientMachine.cpp:85] Initing parameters..
+I0727 05:01:50.690007    25 GradientMachine.cpp:92] Init parameters done.
+```
+[More  optins and cmd aoubt paddle cloud](https://github.com/PaddlePaddle/cloud/blob/develop/doc/usage_cn.md)
````
--> For more information, please refer to [PaddleCloud](https://github.com/PaddlePaddle/cloud/blob/develop/doc/usage_cn.md)

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,41 @@
+**Step3:** Get logs from paddle cloud by cmd: `paddlecloud logs -n 10000 deepspeech20170727130129`.
+
+```
+$ paddlecloud logs -n 10000 deepspeech20170727130129
````
Why to paste such logs?

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Run DS2 by public data
+
+**Step1: ** Make sure current dir is `models/deep_speech_2/cloud/`
```
How to prepare data in cloud for users? Please prepare a script helping users to pack and upload data.

xinghai-sun commented on deep_speech_2/cloud/pcloud_prepare_data.py:
```diff
@@ -0,0 +1,61 @@
+"""
```
rename the file to "update_data.py"

xinghai-sun commented on deep_speech_2/cloud/pcloud_prepare_data.py:
```diff
@@ -0,0 +1,61 @@
+
+if __name__ == '__main__':
+    gen_pcloud_data(args.manifest_path, args.out_tar_path,
+                    args.out_manifest_path)
```
Add a function to upload the data.

wanghaoshuang replied:
How about implementing uploading data in `pcloud_submit.sh`? And rename `pcloud_prepare_data.py` to `pack_data.py` whose duty is just packing data locally?

xinghai-sun replied:
I don't think we have to upload data every time when we submit a training job, so it would be better to separate data uploading from training job submission?

xinghai-sun commented on deep_speech_2/cloud/pcloud_prepare_data.py:
```diff
@@ -0,0 +1,61 @@
+    help="Output tar file path. (default: %(default)s)")
+parser.add_argument(
+    "--out_manifest_path",
+    default="./data/dev.mani",
```
Set default name to `manifest.train.cloud`?

wanghaoshuang replied:
I renamed it to cloud.train.manifest. Use manifest as suffix for indicating the format of this file?

xinghai-sun replied:
No problem. But in other part of the project, we use e.g. `manifest.train`, what about keeping them consistent in this PR and send another PR to modify all?

xinghai-sun commented on deep_speech_2/cloud/pcloud_split_data.py:
```diff
@@ -0,0 +1,50 @@
+    help="Data tar file path. (default: %(default)s)")
+parser.add_argument(
+    "--out_manifest_path",
+    default='./cloud/data/dev.mani.split',
```
--> manifest.train.local ?

wanghaoshuang replied:
I renamed it to  `local.train.manifest`. Use manifest as suffix for indicating the format of this file?

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -215,6 +225,46 @@ def vocab_list(self):
+
+    def _read_soundbytes(self, filepath):
+        """
+        Read bytes from file.
```
merge L240 and L239

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -215,6 +225,46 @@ def vocab_list(self):
         return self._speech_featurizer.vocab_list
 
+    def _parse_tar(self, file):
+        """
```
Merge L228 and L229

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -215,6 +225,46 @@ def vocab_list(self):
+
+    def _process_utterance(self, filename, transcript):
+        """Load, augment, featurize and normalize for speech data."""
+        speech_segment = SpeechSegment.from_bytes(
```
可以复用原来的_process_utterance, 因其filename参数可以接受file object。建议在该函数调用前，增加一个方法例如_filepath_to_object处理一下传入的filename，如果是tar格式的路径，从tar文件提取并返回它的file object。

wanghaoshuang committed:
Implement uploading data in submit scripts and fix issues

wanghaoshuang committed:
Implement uploading data to PaddleCloud
1. Refine data_utils/data.py, reuse process_utterance function.
2. Modified README.
3. Implement uploading data in cloud/upload_data.py
4. Merge branch 'develop' of https://github.com/PaddlePaddle/models into ds2_pcloud

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/models into ds2_pcloud

wanghaoshuang committed:
Merge branch 'ds2_pcloud' of https://github.com/wanghaoshuang/models into ds2_pcloud

wanghaoshuang committed:
remove binary files

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,81 @@
+```
+
+- `TEST_MANIFEST`: Absolute path of train data manifest file in local filesystem.This file has format like  TRAIN_MANIFEST.
+
````
Remove L19.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,81 @@
+": "nor is mister ..."}
+```
+
+- `TEST_MANIFEST`: Absolute path of train data manifest file in local filesystem.This file has format like  TRAIN_MANIFEST.
````
1. `.This file` --> `. This file`. Don't forget a dot mark.
2. TRAIN_MANIFEST --> `TRAIN_MANIFEST`

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+- `TEST_MANIFEST`: Absolute path of train data manifest file in local filesystem.This file has format like  TRAIN_MANIFEST.
+
+- `VOCAB_FILE`:  Absolute path of vocabulary file in local filesytem.
+- `MEAN_STD_FILE`: Absolute path of vocabulary file in local filesytem.
```
vocabulary file --> normalizer's statistic file

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+
+>Note: Make sure current directory is `models/deep_speech_2/cloud/`
+
+## Step1  Configure data set
```
Step1 --> Step-1

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Step1  Configure data set
+
+You can configure your input data and output path in pcloud_submit.sh:
```
--> Configure your input data and output path in `pcloud_submit.sh`:

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,81 @@
+
+You can configure computation resource in pcloud_submit.sh:
+```
+# Configure computation resource and submit job to PaddleCloud
````
Remove L32

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+
+>Note: Upload will be skipped if target file has existed in  ${CLOUD_DATA_DIR}.
+
+## Step2  Configure computation resource
```
Step2 --> Step-2

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Step2  Configure computation resource
+
+You can configure computation resource in pcloud_submit.sh:
```
--> Configure ... in `pcloud_submit.sh`

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+﻿# Run DS2 on PaddleCloud
+
```
Add a Prerequisite session to remind users to install paddlecloud client?

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
```diff
@@ -0,0 +1,81 @@
+- `CLOUD_MODEL_DIR`: Absolute path in PaddleCloud filesystem. PaddleCloud trainer will save model to this directory.
+
+
+>Note: Upload will be skipped if target file has existed in  ${CLOUD_DATA_DIR}.
```
${CLOUD_DATA_DIR} --> `CLOUD_DATA_DIR`

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/README.md:
````diff
@@ -0,0 +1,81 @@
+ -entry "sh pcloud_train.sh ${CLOUD_DATA_DIR} ${CLOUD_MODEL_DIR}" \
+ ${DS2_PATH}
+```
+For more information, please refer to[PaddleCloud](https://github.com/PaddlePaddle/cloud/blob/develop/doc/usage_cn.md#提交任务)
````
"to[PaddleCloud]" --> "to [PaddleCloud]"

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/pcloud_train.sh:
```diff
@@ -0,0 +1,31 @@
+
+python train.py \
+--use_gpu=1 \
+--trainer_count=4 \
```
Please set the default values to those listed in train.py.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/split_data.py:
```diff
@@ -0,0 +1,50 @@
+"""
+This tool is used for splitting data into each node of
```
Merge L1 and L2.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+"""
+This tool is used for preparing data for DeepSpeech2 trainning on paddle cloud.
```
Merge L1 and L2

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+Finally, we will get a tar file and a manifest with sound file name, duration
+and text.
+"""
+import json
```
Add 
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

```
to all newly added files.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+parser.add_argument(
+    "--cloud_data_path",
+    required=True,
+    default="",
```
If `required = True`, can we remove `default=""`?

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+
+def pack_data(manifest_path, out_tar_path, out_manifest_path):
+    '''
+    1. According manifest, tar sound files into out_tar_path
```
--> """ xxxxx
--> According to the manifest,

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+
+
+if __name__ == '__main__':
+    cloud_train_manifest = "%s/%s" % (args.cloud_data_path, TRAIN_MANIFEST)
```
Use `os.path.join`

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+    ret = 1
+    # train data
+    if args.train_manifest_path != "":
+        ret = call(['paddlecloud', 'ls', cloud_train_manifest])
```
Should we need `try except` ?

wanghaoshuang replied:
这里不会因为cloud文件的不存在为而抛异常，没必要`try except`。如果是 `PaddleCloud not installed`等原因带来的异常，我们应该让它快速失败，然后在调用这个脚本的地方处理异常。

xinghai-sun commented on deep_speech_2/cloud/upload_data.py:
```diff
@@ -0,0 +1,147 @@
+                      local_test_manifest)
+            call(
+                ['paddlecloud', 'cp', local_test_manifest, cloud_test_manifest])
+            call(['paddlecloud', 'cp', local_test_tar, cloud_test_tar])
```
Write a function to do cloud fs cp.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -82,12 +89,15 @@ def __init__(self,
         self._rng = random.Random(random_seed)
         self._epoch = 0
+        # for caching tar files info
+        self.tar2info = {}
```
remove L93-94, not used.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -215,6 +225,37 @@ def vocab_list(self):
+        return f, result
+
+    def _get_file_object(self, file):
+        """Get file object by file path.
```
Add a blank line after L239.

wanghaoshuang replied:
Fixed.

xinghai-sun commented on deep_speech_2/data_utils/data.py:
```diff
@@ -215,6 +225,37 @@ def vocab_list(self):
+            return local_data.tar2object[tarpath].extractfile(
+                local_data.tar2info[tarpath][filename])
+        else:
+            return open(file)
```
open(file, 'r')

wanghaoshuang committed:
Fix some syntax errors.

wanghaoshuang committed:
Change the default values in pcloud_train to those listed in train.py.
